feat: Add Python pipelines to consume Chainguard Libraries

EyeCantCU opened the pull request:
This adds a few pipelines for both authentication with our indexes, and for building packages that consume Chainguard Libraries:

- **py/bundle**: Leverages uv to install Python packages from Guarded Libraries. Used exclusively for bundles of Python packages
- **auth/uv**: Configures uv to pull Chainguard Libraries
- **auth/pip**: Configures pip to pull Chainguard Libraries

A few packages have been moved over to Chainguard Libraries in the process:

- **jupyter-base-notebook**: Demonstrates packaging of a Python bundle, uses some libraries. Some, such as notebook itself, were not present at the time of testing
- **superset-5.0**: Demonstrates working authentication with pip in a virtual environment, consumes some libraries. Will need to be fully migrated in the future

This change also globally disallows downloading external builds of Python with uv. External Python builds should never be used to produce packages.

EyeCantCU committed:
chore(pipelines/auth): Configure pip to retrieve guarded libraries

Introduces a pipeline that configures the index url for pip with auth
for guarded libraries

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
feat: Add new uv-backed pipelines for Python

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(auth/pip): Add toggle to disable external index fallback

Adds a 'guarded' toggle that enforces use of Chainguard Libraries by
default. Explicit opt-out is required to fallback to non-Chainguard
Libraries

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(auth/uv): Add toggle to disable external index fallback

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(pipelines/py): Remove venv pipeline

Not needed anymore. We'll install directly to system site packages
instead

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
world: Do not download external Python builds with UV

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(py/install): Various fixes to make this actually work

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(py3-setuptools): Migrate to guarded libraries

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(py3-poetry): Use guarded libraries

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(py/install): Support installation of additional dependencies

Some of our packages are bundles that vendor non-direct dependencies.
Support their installation

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(jupyter-base-notebook): Use guarded libraries where we can

Not all dependencies are available via guarded libraries (i.e.
notebook). So use what we can

And minor cleanup. The CVE remediations weren't actually remediating any
CVEs as they were bumping dependencies in the builder itself and not in
the package

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(superset-4.0): Demo guarded libraries auth with pip

This doesn't fully move superset to guarded libraries, more so for
proving auth with pip

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(auth/uv/pip): Correctly retrieve ID token during EBs

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(superset-5.0): Fallback to other indexes

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(jupyter): Bump to Python 3.13

I have no idea how co-installation of both of these packages ever
worked, but fingers crossed

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(jupyter-k8s-hub): Remove unused buildtime dependencies

Python is never used at all

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

justinvreeland commented on jupyterhub-k8s-hub.yaml:
```diff
@@ -1,46 +1,45 @@
-      - py3-sqlalchemy-cockroachdb
-      - py3-statsd
-      - py3-tornado
-      - python3
```
Oh wow that's a huge improvement versioning them.

justinvreeland commented on pipelines/auth/pip.yaml:
```diff
@@ -0,0 +1,53 @@
+      fi
+
+      echo "Creating pip config..."
+      cat > $HOME/.config/pip/pip.conf <<EOF
```
Double quote this and 35 please. I don't think we run any shellcheck on pipelines unfortunatly.

Taffer replied:
Hooking `shellcheck` up in your editor is always a good idea.

Taffer replied:
We have patch files as stand-alone files in `{package}/`; what about shell scripts? Or is there a strong preference for keeping them inline in the `{package}.yaml`?

justinvreeland reviewed (changes requested):
Looks good, looking forward to trying it out.

tulilirockz commented:
+1 here, I _really_ want this

EyeCantCU commented on py3-poetry.yaml:
```diff
@@ -25,19 +25,12 @@ data:
-      - ca-certificates-bundle
       - py3-supported-pip
       - py3-supported-poetry-core
-      - wolfi-base
```
It seems unnecessary to include it if it isn't used directly at build time

EyeCantCU committed:
chore(pipelines/auth): Make shellcheck happy

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore: Use python-version, fix epochs

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(setuptools/poetry): Restore git-checkout

Useful for populating buildtime SBOM

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(jupyter-base-notebook): Do not install k8s hub during tests

We do not have any images that consume both of these concurrently, and
what previously allowed them to work together at all is flawed

Following history, these packages were never meant to be co-installed at
all

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(jupyter-base-notebook): Update to v7.4.7

And add git-checkout so that we populate the buildtime SBOM

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore: Cleanup buildtime deps

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(jupyter-base-notebook): pip -> uv

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(jupyterhub-k8s-hub): Restore jupyterhub binaries at runtime

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore: Revert changes to setuptools and poetry

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(pipelines/py): Only support bundles for now

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
yam it

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

justinvreeland commented:
It looks like Jupyter lost all of it's cmd provides:

```
- 	Provides: []string{
- 		"cmd:alembic=7.4.6-r0", "cmd:certipy=7.4.6-r0", "cmd:debugpy-adapter=7.4.6-r0",
- 		"cmd:debugpy=7.4.6-r0", "cmd:httpx=7.4.6-r0", "cmd:ipython3=7.4.6-r0",
- 		"cmd:ipython=7.4.6-r0", "cmd:jlpm=7.4.6-r0", "cmd:jsonpointer=7.4.6-r0",
- 		"cmd:jsonschema=7.4.6-r0", "cmd:jupyter-dejavu=7.4.6-r0",
- 		"cmd:jupyter-events=7.4.6-r0", "cmd:jupyter-execute=7.4.6-r0",
- 		"cmd:jupyter-kernel=7.4.6-r0", "cmd:jupyter-kernelspec=7.4.6-r0",
- 		"cmd:jupyter-lab=7.4.6-r0", ...,
- 	},
```

EyeCantCU commented:
@justinvreeland I intentionally enabled no-provides as the package is a bundle of multiple Python packages. We have other packages that provide those commands as well

A caller of `cmd:jupyterhub` likely wants `py3.13-jupyterhub-bin` and not `jupyter-base-notebook`

I also incremented Python to 3.13 explaining the library changes

EyeCantCU commented:
`libm`, `librt`, and `libdl` are curious... looking into those

jamonation commented:
Is it worth splitting out the package changes so that the auth changes can be used here separately?

EyeCantCU commented:
It doesn't. Let's wait until we've dug into this more and understand what's going on

justinvreeland commented:
> All changes are related to us building against newer glibc. Let's proceed
Awesome thanks for looking into that!

dannf commented on pipelines/auth/pip.yaml:
```diff
@@ -0,0 +1,53 @@
+name: Configure pip to fetch Chainguard Libraries dependencies.
```
Do we need separate pipelines for pip and uv, or could we just configure them both in the same pipeline?

EyeCantCU replied:
We cam throw them in the same pipeline

dannf commented on pipelines/auth/pip.yaml:
```diff
@@ -0,0 +1,53 @@
+      fi
+
+      echo "Creating pip config..."
+      cat > "$HOME/.config/pip/pip.conf" <<EOF
```
Should we error out if this file already exists (same q for uv)?

dannf commented on pipelines/auth/pip.yaml:
```diff
@@ -0,0 +1,53 @@
+  group:
+    description: the group to use for Chainguard Libraries
+    default: "720909c9f5279097d847ad02a2f24ba8f59de36a"
+  guarded:
```
how about "no-index-fallback" or "guarded-only"?

dannf replied:
just the name. The default should be to *only* install guarded things. I like names that i can read and I immediately understand broadly what it does. `guarded` true/false doesn't meet that bar for me. We're installing guarded libs either way. "only-guarded" or "no-index-fallback" or something like that tells me what it does on the tin.

TBH, I'm actually not even sure I like having a way to enable a fallback. I'm sure we'll need an escape hatch at times, but I'd prefer something more granular and declarative than "whatever the resolver happens to choose at that moment". I'd like for developers to explicitly list the things that they are pulling from any other index and why.

WDYT about putting this on hold? Or making it a flag that is clearly only meant for debug?

dannf commented on pipelines/auth/pip.yaml:
```diff
@@ -0,0 +1,53 @@
+      fi
+
+      echo "Removing pip config, if exists..."
+      rm -rf "$HOME/.config/pip" && mkdir -p "$HOME/.config/pip"
```
Should we error here instead of continuing? That would make debugging easier, in case some package was relying on an existing configuration here.

dannf commented on pipelines/py/bundle.yaml:
```diff
@@ -0,0 +1,36 @@
+
+inputs:
+  dependencies:
+    description: "Indirect dependencies to install."
```
It's not clear to me what this means, or what syntax to use to specify them.

EyeCantCU replied:
We have some packages like jupyter base notebook that install indirect dependencies. This allows you to pass a list of indirect deps

dannf commented on pipelines/py/bundle.yaml:
```diff
@@ -0,0 +1,36 @@
+    description: "Indirect dependencies to install."
+    default: ""
+  package:
+    description: "Name of packages to install."
```
I expect we'd normally want the full pin syntax. What if we allowed passing in a requirements file and/or lockfile instead?

dannf replied:
Yeah, I suppose ultimately we want to support both. It would be annoying to have to generate a requirements file if you just need a single package. But there are a number of packages that will benefit from `requirements.txt` support.

dannf commented on pipelines/py/bundle.yaml:
```diff
@@ -0,0 +1,36 @@
+    description: "Name of packages to install."
+    required: true
+  python-version:
+    description: "Python version to use."
```
How about a python path instead? Looks like `UV_PYTHON` supports that. It would be more consistent with our other python pipelines, and would lend itself to venvs in cases where that's necessary.

dannf reviewed (changes requested):
Thanks @EyeCantCU! some comments inline

EyeCantCU committed:
feat(pipelines/py): Add pipeline for generating locked requirements

Uses uv to initialize a project, add all packages, and then exports a
lockfile

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
feat(py/bundle): Install from requirements file

Use a requirements file instead of passing dependencies directly to the
pipeline

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(jupyter-base-notebook): Add and use locked requirements

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(compile-requirements): Add main package to project last

Ensure that the main package dictates final requirements

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(auth/python): guarded -> no-index-fallback

Give toggle for enabling external indexes clearer naming

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(py/bundle): Add initial virtual environment support

This adds support for packaging things as virtual environments. While
bundles should be the preferred default, providing a path via the venv
opt will create a virtual environment at the given path instead of
bundling everything in system site packages

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(py/bundle): Install dependencies in the build env if required

--reinstall force ignores any dependencies that already exist in the
build environment and installs them regardless. I can't think of many
situations where this would happen, but its a possibility we can
mitigate by doing this

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
feat(pipelines/py): Pipeline for packaging individual Python packages

This adds a pipeline for packaging individual dependencies. While
bundles should usually always be preferred, this offers a path forward
for individual packages if ever needed

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(py3-django): Use Guarded Libraries

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
feat(py/compile-requirements): Support pyprojects and dependency groups

This completely overhauls the pipeline to support pyprojects and
dependency groups. Not all uv export options are covered becuase I don't
ever really see us using them (--all-extras, --no-extra, etc) and there
functionality is still intact but in a different way (define extra
dependency groups explicitly)

The behavior can always be changed later if we see the need to

Otherwise, this also adds a way to constrain Python versions recognized
as supported by uv... I don't like this at all, but I realized working
on jupyter notebook that if upstream hasn't set constraints properly
then there really isn't much else we can do as a workaround aside from
vendoring a patch

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

smoser reviewed (commented):
general comment. 
this is a lot of yaml bash.

tw provides one way of getting out of yaml-bash (even if only replacing it with bash).  It also provides a solution to copying pipelines around.
